### PR TITLE
Feature: Leave Management Challenge

### PR DESCRIPTION
**Pull request by mhafiziskandar**
## Summary

Built the leave management API service layer for the FastAPI scaffold, including leave creation, approval/rejection, cancellation, filtered listing with pagination, and leave balance retrieval.

The implementation covers the required business rules such as:
- date validation
- overlapping leave prevention
- insufficient balance checks
- manager-only approval
- self-approval prevention
- balance deduction on approval and restoration on cancellation

I also added comprehensive unit and integration tests for the core workflows and edge cases.

## Design Document

Attached in this PR as `DESIGN.md`.

Key points covered in the design doc:
- API design decisions
- data model considerations
- edge cases and tradeoffs
- concurrency considerations
- what I would improve with more time

## Implementation Notes

A few implementation decisions worth calling out:

- Leave balance is deducted on approval rather than at request creation. This keeps pending requests from prematurely consuming quota.
- Balance is re-checked at approval time to handle the gap between request creation and manager review.
- Overlap detection uses an inclusive date-range intersection check.
- Unpaid leave does not consume balance.
- If no balance row exists for a leave type/year, the request is treated as having zero available days rather than creating a zero-day balance record during failed validation.
- I used offset-based pagination for simplicity and readability in this challenge.
- Concurrency is discussed in `DESIGN.md`; the implementation uses `with_for_update()` on balance reads, with the caveat that SQLite has limited locking semantics compared to PostgreSQL.
- For local verification, I aligned the project workflow to the repo virtualenv and validated the app using `make test` and `make run`, which wrap the project pytest and uvicorn commands.

## Checklist

- [x] I have written and attached a DESIGN.md
- [x] I have implemented the required service functions in src/services.py
- [x] Tests pass (`make test`, which runs the project venv pytest command)
- [x] Server starts and responds to requests (`make run`, which starts `uvicorn src.app:app --reload`)
- [x] I have considered edge cases (overlapping leave, insufficient balance, self-approval)

## About You

I’m a full stack developer with 5+ years of experience building web applications, backend APIs, system integrations, and workflow-driven platforms. My work has covered HR-related systems, government and regulatory platforms, and internal business tools, which is part of why this challenge felt genuinely relevant to me.

I’m most comfortable working with Node.js, Laravel/PHP, React, MySQL, PostgreSQL, REST APIs, and Git. I’ve also worked with workflow automation using n8n and on systems where secure, reliable integration between services is important.

What I look for in a workplace is a team that values practical engineering, clear communication, and thoughtful product decisions. I enjoy collaborating on backend architecture, building reliable APIs, and working on systems that solve real operational problems for users.

- LinkedIn: [Hafiz Iskandar](https://www.linkedin.com/in/hafiziskandar/)  
- Website: [mhafiziskandar.dev](https://mhafiziskandar.dev/)  
- Email: mhafiziskandar.dev@gmail.com

### CODE REVIEW

File: src/services.py [hunk 1]
```diff
@@ -50,9 +85,59 @@ def create_leave_request(
     - start_date >= today (no back-dating)
     - No overlapping leave requests for the same employee
     - Employee has sufficient leave balance for the requested type
-    - end_date - start_date >= 0 (at least 1 day — or handle half-day logic)
+    - end_date - start_date >= 0 (at least 1 day - or handle half-day logic)
     """
-    raise NotImplementedError("Candidate must implement this")
+    _get_employee_or_404(db, employee_id)
```
Comment (KenWang-Kakitangan) [inline src/services.py]: no variable to hold the returned emp obj.

File: src/services.py [hunk 2]
```diff
@@ -50,9 +85,59 @@ def create_leave_request(
...
+
+    today = date.today()
+    if start_date < today:
+        raise LeaveError("Leave start date cannot be in the past")
+    if end_date < start_date:
```
Comment (KenWang-Kakitangan) [inline src/services.py:95]: end_date could be same as start_date

File: src/services.py [hunk 3]
```diff
@@ -50,9 +85,59 @@ def create_leave_request(
...
+
+    # Overlap check: any active (non-cancelled, non-rejected) request whose
+    # date range intersects [start_date, end_date].
+    overlap = (
+        db.query(LeaveRequest)
+        .filter(
+            LeaveRequest.employee_id == employee_id,
+            LeaveRequest.status.not_in([LeaveStatus.CANCELLED, LeaveStatus.REJECTED]),
+            LeaveRequest.start_date <= end_date,
+            LeaveRequest.end_date >= start_date,
```
Comment (KenWang-Kakitangan) [inline src/services.py:106]: This filter condition with start_date, end_date is not fully correct

File: src/services.py [hunk 4]
```diff
@@ -112,7 +286,17 @@ def get_leave_balances(
...
     """
-    raise NotImplementedError("Candidate must implement this")
+    if year is None:
+        year = date.today().year
+
+    return (
+        db.query(LeaveBalance)
+        .filter(
+            LeaveBalance.employee_id == employee_id,
+            LeaveBalance.year == year,
```
Comment (KenWang-Kakitangan) [inline src/services.py:287]: it is not enough to get a leave balance without leave type info.

File: src/services.py [hunk 5]
```diff
@@ -84,7 +205,40 @@ def cancel_leave_request(
...
+    if lr.status == LeaveStatus.REJECTED:
+        raise LeaveError("Rejected leave requests cannot be cancelled")
+
+    was_approved = lr.status == LeaveStatus.APPROVED
+
+    lr.status = LeaveStatus.CANCELLED
+
+    if was_approved and lr.leave_type != LeaveType.UNPAID:
+        year = lr.start_date.year
+        balance = (
```
Comment (KenWang-Kakitangan) [inline src/services.py]: why not use _get_balance function.

File: src/services.py [hunk 6]
```diff
@@ -34,6 +34,41 @@ class CannotModifyApprovedLeaveError(LeaveError):
...
+    return lr
+
+
+def _count_working_days(start_date: date, end_date: date) -> float:
+    """Count calendar days inclusive. Weekend-skipping left as a future enhancement
+    (see DESIGN.md) - we use calendar days to keep the model simple and predictable."""
+    return float((end_date - start_date).days + 1)
+
+
+def _get_balance(
```
Comment (KenWang-Kakitangan) [inline src/services.py:57]: there is another function get_leave_balances at line 281.

Comment (mhafiziskandar) [reply src/services.py:57]: The get_leave_balances on line 281 is being used intentionally for lock-free since it's for a read-only listing endpoint. The _get_balance got with_for_update() which is only needed because it's being used in the create_leave_request and approve_leave_request, where we used to read and then deduct/restore the balance. Also to prevent concurrent over deduction.

File: src/services.py (hunk 1 above)
Comment (mhafiziskandar) [reply src/services.py]: It was intended as a validation guard clause that will validates existence of the employee otherwise it will raises the 404. I did a fix in commit 31db33d to captured it as employee for clarity and to match the pattern that I used in the approve_leave_request.

File: src/services.py (hunk 2 above)
Comment (mhafiziskandar) [reply src/services.py:95]: Yes, that's my intention for the single day leave where end_date == start_date. The checks only rejects end_date < start_date which equal dates pass through should be fine.

File: src/services.py (hunk 3 above)
Comment (mhafiziskandar) [reply src/services.py:106]: Thank you Ken, but can you clarify what's the incorrect here. Is it the boundary math itself ? Or do you mean by this filter doesnt capture the scope by leave_type or did you mean by something else ?

File: src/services.py (hunk 5 above)
Comment (mhafiziskandar) [reply src/services.py]: It was an oversight on my site. When I wrote the cancel logic I did built the balance lookup inline instead of reusing that function which already does the exact same query. I did a fix on commit 5c24880

File: src/services.py (hunk 4 above)
Comment (mhafiziskandar) [reply src/services.py:287]: Thank you Ken. I try to double check the leave type as you mention, the LeaveBalance model already includes that in every row returned here so when I return the full ORM object it carries it along. I also ran it locally to confirm and I get this kind of response 

get_leave_balances(db, employee_id=2) 
returned 2 row(s):
id=1 employee_id=2 leave_type=LeaveType.ANNUAL year=2026 total_days=14.0 used_days=0.0 remaining_days=14.0
id=2 employee_id=2 leave_type=LeaveType.SICK year=2026 total_days=12.0 used_days=0.0 remaining_days=12.0

The query intentionally didnt filter the leave_type so it can return all the balance types for that employee in one call. But let me know if you have any specific response where it looked missing. I would be happy to check on it further
